arrow-fx-stm/src/main/kotlin/arrow/fx/stm/STM.kt

```diff
@@ -208,22 +209,90 @@ interface STM {
    * Rerun the current transaction.
    *
    * Aborts the transaction and suspends until any of the accessed [TVar]'s changed, after which the transaction restarts.
```

Comment:
When `retry` is called it first validates the transaction to see if it can restart right now. If not it registers itself at any accessed `TVar` (checking for changes before and after registering to not miss updates) and then it suspends until its `Continuation` is called from any changed `TVar`.

Also `TVar` is the only primitive datastructure, all others are built upon it.

Reply:
> Also TVar is the only primitive datastructure, all others are built upon it.
Yes, I was just wondering if we should repeat that here. Something like:

> Aborts the transaction and suspends until any of the accessed [TVar]'s changed, so all data structures found in the STM package , after which the transaction restarts.

I think the explanation you shared here on `retry` also more clear IMO!

Reply:
>> Also TVar is the only primitive datastructure, all others are built upon it.

> Yes, I was just wondering if we should repeat that here. Something like:

> > Aborts the transaction and suspends until any of the accessed [TVar]'s changed, so all data structures found in the STM package , after which the transaction restarts.

This is pretty similar to how GHC documents it, but I can see this being confusing and your addition does clear that up :+1: 

> I think the explanation you shared here on retry also more clear IMO!

It is, but imo it goes too much into implementation detail which I'd like to avoid. Not sure how to phrase it better :/